src/theory/arrays/theory_arrays.cpp

```diff
@@ -1285,36 +1303,92 @@ void TheoryArrays::check(Effort e) {
+            if (d_conflict) { Debug("pf::array") << "Entering the skolemization branch" << std::endl; }
+
             NodeManager* nm = NodeManager::currentNM();
             TypeNode indexType = fact[0][0].getType()[0];
-            TNode k = getSkolem(fact,"array_ext_index", indexType, "an extensional lemma index variable from the theory of arrays", false);
-
-            Node ak = nm->mkNode(kind::SELECT, fact[0][0], k);
-            Node bk = nm->mkNode(kind::SELECT, fact[0][1], k);
-            Node eq = ak.eqNode(bk);
-            Node lemma = fact[0].orNode(eq.notNode());
-            if (options::arraysPropagate() > 0 && d_equalityEngine.hasTerm(ak) && d_equalityEngine.hasTerm(bk)) {
-              // Propagate witness disequality - might produce a conflict
-              d_permRef.push_back(lemma);
-              d_equalityEngine.assertEquality(eq, false, lemma, eq::MERGED_ARRAYS_EXT);
-              ++d_numProp;
+
+            if (!d_proofsEnabled) {
+              Debug("pf::array") << "Check: kind::NOT: array theory making a skolem" << std::endl;
+
+              TNode k = getSkolem(fact,"array_ext_index", indexType, "an extensional lemma index variable from the theory of arrays", false);
+
+              if (options::proof()) {
+                ProofManager::getSkolemizationManager()->registerSkolem(fact, k);
+              }
+
+              Node ak = nm->mkNode(kind::SELECT, fact[0][0], k);
+              Node bk = nm->mkNode(kind::SELECT, fact[0][1], k);
+
+              Node eq;
+              eq = ak.eqNode(bk);
+              Node lemma = fact[0].orNode(eq.notNode());
+
+              if (options::arraysPropagate() > 0 && d_equalityEngine.hasTerm(ak) && d_equalityEngine.hasTerm(bk)) {
+                // Propagate witness disequality - might produce a conflict
+                d_permRef.push_back(lemma);
+                d_equalityEngine.assertEquality(eq, false, fact, eq::MERGED_ARRAYS_EXT);
+                ++d_numProp;
+              }
+
+              Trace("arrays-lem")<<"Arrays::addExtLemma " << lemma <<"\n";
+              d_out->lemma(lemma);
+              ++d_numExt;
+            } else {
+              Debug("pf::array") << "In Array's Check (proof phase), handling fact = " << fact << std::endl;
+
+              TNode k;
+              if (ProofManager::getSkolemizationManager()->hasSkolem(fact)) {
+                k = ProofManager::getSkolemizationManager()->getSkolem(fact);
+                Debug("pf::array") << "Skolem = " << k << std::endl;
+
+                Node ak = nm->mkNode(kind::SELECT, fact[0][0], k);
+                Node bk = nm->mkNode(kind::SELECT, fact[0][1], k);
+                Node eq = ak.eqNode(bk);
+                Node lemma = fact[0].orNode(eq.notNode());
+
+                // Only in proof mode, if the terms ak and bk are not yet registered, we'd like to register them.
+                if (!d_equalityEngine.hasTerm(ak)) {
+                  preRegisterTermInternal(ak);
```

Comment:
I've cleaned this code and added comments to highlight the differences between the solution and replay passes (they don't appear very great, now).
Regarding the pre-registration, it might be possible to drop, but I'll need to run some regression tests to make sure.